### ISSUE TITLE
Insert images in memory, rather than from the filesystem

### ISSUE THREAD

**Issue by epalm** [reingart/pyfpdf#85]
I'm first using Pillow to generate an image, and then inserting the image into a pdf using FPDF.

Currently I have to
- create the image using Pillow
- save the image to disk
- create a pdf with FPDF
- insert the image into the pdf using `image('path/to/file', ...)` which calls `_parsepng(filename)` which calls `load_resource(filename)` which finally calls `open(filename, "rb")`
- delete the image from disk

Since I have the image already loaded in memory, if I could pass those bytes to the `image` function then I could avoid touching the disk altogether.

**Comment by RomanKharin** [comment]
Thanks to you too.
There are two assumtions:
1) load_resource() are considered to be overloaded for other data sources
2) reading files directly from memory are handfull too, but i can imagine simple and clean method for specifying this, other than 'base64://' pseudo-url

**Comment by epalm** [comment]
Right.  I would think that the `load_resource` function would remain untouched (there's nothing to load, you already have the bytes), and the `_parsexxx` functions could just proceed without calling `load_resource` at all.

**Comment by alhman** [comment]
same here, many thanks for this great library.

I'm having exactly the same issue.  It would be great if you could provide a method for specifying the image in memory.

At the moment I would also be fine with a `load_resource` workaround. I have a dependency to fpdf for my project but it seems the package on PyPI does not contain the latest version. In this version there is no `load_resource` method. Could you update the version on PyPI?

**Comment by alexanderankin** [comment]
Under the assumption that the Pillow people have figured out the technical kinks that previously prevented people from using their library:

check out this implementation - i tried my best but it came out slower than the original: [usage](https://github.com/alexanderankin/pyfpdf/blob/image-attempt-1/test/image/png_images/png_image_test.py), and partial implementation details:
* [def _parse_png](https://github.com/alexanderankin/pyfpdf/blob/image-attempt-1/fpdf/fpdf.py#L2060). This still reads through the bits in memory, bit by bit. Perhaps better rewrite the whole thing to pick stuff out from the data that PIL automatically?
* [def image](https://github.com/alexanderankin/pyfpdf/blob/image-attempt-1/fpdf/fpdf.py#L1140)

Looking forward to your guys feedback, if this is something i should finish for jpg file and merge back into at least my own master.

for reference, you can run that test either by changing directory into that folder and running the file, (this will use the timing code i have in the file), or by checking out the branch and simply running `tox` command. This version of the library has [explicit Pillow dependency](https://github.com/alexanderankin/pyfpdf/blob/image-attempt-1/setup.py#L39).

**Comment by alexanderankin** [comment]
reading off disk isn't too much slower on my machine, might be the ssd:

```java
$ python png_image_test.py -vv
test_insert_png_files (__main__.InsertPNGSuiteFiles) ... ok

__main__.InsertPNGSuiteFiles: 0.489
test_insert_png_memory (__main__.InsertPNGSuiteMemory) ... ok

__main__.InsertPNGSuiteMemory: 0.544

----------------------------------------------------------------------
Ran 2 tests in 1.033s

OK
```

**Comment by alexanderankin** [comment]
The numbers seem to have righted themselves after testing out on a clean virtual machine install after some big fixes in the test.
```java
$ python png_image_test.py -vv
test_insert_png_files (__main__.InsertPNGSuiteFiles) ... ok

__main__.InsertPNGSuiteFiles: 0.836
test_insert_png_memory (__main__.InsertPNGSuiteMemory) ... ok

__main__.InsertPNGSuiteMemory: 0.738

----------------------------------------------------------------------
Ran 2 tests in 1.575s

OK
```

**Comment by pablote** [comment]
Just found this through Google. Is adding images from memory supported now? going through files is not really an option for me. Thanks.

**Comment by alexanderankin** [comment]
> Just found this through Google. Is adding images from memory supported now? going through files is not really an option for me. Thanks.

https://github.com/alexanderankin/pyfpdf/blob/image-attempt-1/test/image/png_images/png_image_test.py

**Comment by BlockChainCaffe** [comment]
Same here
Please add a funcion. Thanks

**Comment by alexanderankin** [comment]
i have been releasing stuff on pyfpdf2, my fork of this project, you can
pass BytesIO instance as name of the image and it uses pillow

**Comment by Lucas-C** [comment]
PyFPDF is not maintained anymore, you may want to check [PyFPDF/fpdf2](https://github.com/PyFPDF/fpdf2) as its successor, with a 99%-compatible API

It **does** support inserting images from memory, as bytes
